sunpy/net/dataretriever/sources/noaa.py

```diff
@@ -148,37 +151,79 @@ class SRSClient(GenericClient):
-        for day in all_dates:
-            end_year = int(day.end.strftime('%Y'))
-            if end_year > today_year or end_year < 1996:
-                continue
-            elif end_year == today_year:
-                suffix = '{}/SRS/{}SRS.txt'.format(
-                    end_year, day.end.strftime('%Y%m%d'))
-            else:
-                suffix = '{}/{}_SRS.tar.gz'.format(
-                    end_year, day.end.strftime('%Y'))
-            url = base_url + suffix
-            result.append(url)
+
+        # Validate time range and return early if out side min/max
+        cur_year = Time.now().datetime.year
+        req_start_year = timerange.start.datetime.year
+        req_end_year = timerange.end.datetime.year
+        start_year = req_start_year if self.MIN_YEAR < req_start_year <= cur_year else None
+        end_year = req_end_year if self.MIN_YEAR < req_end_year <= cur_year else None
+        if start_year is None and end_year is None:
+            return result
+
+        # Search for tarballs for all years in the query
+        if end_year == start_year:
+            end_year += 1
+        tarball_timerange = TimeRange(f'{start_year}-01-01', f'{end_year}-01-01')
+        tarball_urls = dict()
+        tarball_scraper = Scraper(self.BASE_URL + '%Y/%Y_SRS.tar.gz')
+        tarballs = tarball_scraper.filelist(tarball_timerange)
+        max_tarball_year = None
+        for tb_url in tarballs:
+            date = tarball_scraper._extractDateURL(tb_url)
+            year = date.to_datetime().year
+            max_tarball_year = year
+            tarball_urls[year] = tb_url
+            log.debug('Tarball found for year %d', year)
+
+        # Create a new time range for the times not covered by tarballs, have to assume tarballs
+        # cover a year, and look for individual srs file for this time range.
+        srs_urls = dict()
+        min_file_year = max_tarball_year if max_tarball_year else start_year
+        min_file_date = datetime(min_file_year+1, 1, 1, 0, 0, 0)
+        if min_file_date <= timerange.end.datetime:
+            file_timerange = TimeRange(f'{min_file_year}-01-01', timerange.end)
+            srsfile_scraper = Scraper(self.BASE_URL + '%Y/SRS/%Y%m%dSRS.txt')
+            srsfiles = srsfile_scraper.filelist(file_timerange)
+            for srs_url in srsfiles:
+                date = srsfile_scraper._extractDateURL(srs_url)
+                srs_urls[(date.datetime.year, date.datetime.month, date.datetime.day)] = srs_url
+                log.debug('SRS file found for year %d', date)
+
+        # Now iterate over all days and if the day is in a year we have a tarball for or a day there
+        # is a individual srs file add to the result with corresponding extdict
+        for day in timerange.get_dates():
+            day_ymd = (day.datetime.year, day.datetime.month, day.datetime.day)
+            extdict = {name: getattr(day.datetime, name) for name in ['year', 'month', 'day']}
```

Comment:
Would this work?
```suggestion
            extdict = {name: segment for segment in day_ymd}
```

Reply:
No as name isn't defined then, I could make `day_ymd` a dictionary to start but the lookups in the loop get messy